contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/__init__.py

```diff
@@ -12,31 +12,89 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import threading
+import time
 
 from opencensus.ext.azure.common import Options
+from opencensus.metrics.export import value
+from opencensus.metrics.export.metric_descriptor import MetricDescriptorType
 from opencensus.metrics import transport
 from opencensus.stats import stats
+from opencensus.ext.azure.common import utils
+from opencensus.ext.azure.common.protocol import Data
+from opencensus.ext.azure.common.protocol import DataPoint
+from opencensus.ext.azure.common.protocol import Envelope
+from opencensus.ext.azure.common.protocol import MetricData
+from opencensus.ext.azure.common.transport import TransportMixin
 
 __all__ = ['MetricsExporter', 'new_metrics_exporter']
 
 
-class MetricsExporter(object):
+class MetricsExporter(TransportMixin):
     """Metrics exporter for Microsoft Azure Monitor."""
 
-    def __init__(self, options):
+    def __init__(self, options=None):
+        if options is None:
+            options = Options()
         self.options = options
         if not self.options.instrumentation_key:
             raise ValueError('The instrumentation_key is not provided.')
-        self._md_cache = {}
-        self._md_lock = threading.Lock()
 
     def export_metrics(self, metrics):
-        metrics = list(metrics)
+        if not metrics:
+            return
+        envelopes = []
         for metric in metrics:
-            # TODO: implement export to Azure Monitor functionality
-            print(repr(metric))
+            # No support for histogram aggregations
+            if metric.descriptor.type == MetricDescriptorType.CUMULATIVE_DISTRIBUTION:
+                continue
+            envelopes.append(self.metric_to_envelope(metric))
+            self._transmit(envelopes)
```

Comment:
What if transmission failed?

Reply:
How to control the max batch size (number of items) that we export?

Reply:
Since we do not have a defined retry logic for Metrics yet, the exporter simply ignores the transmissions that have failed and logs an error message with the given status code. I do not want to throw and exception and stop the application if a single transmission fails.

Reply:
We have not yet defined what "number of items" means in terms of metrics (# of metrics, time series or points?). The injestion service also only accepts a single metric at a time, the list passed in is simply for convention (so match with stats design). I say we can keep it as it is for now (sends all time series passed in) until we can define what this means.

Reply:
`_transmit` has [dependency on the storage](https://github.com/census-instrumentation/opencensus-python/blob/master/contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py#L106), we need to decouple it (or make storage optional if we don't use it in metrics exporter).

Reply:
Yes you are correct. Leaving the logic to run into the exception is not a good practice. I fixed it so that it will only pass to storage if storage is being used (not being used for Metrics exporter). This means that for now, any partial success response we get back from the ingestion service will be discarded.

Reply:
The ingestion service only accepts a single data point at a time, so there is no need for batching.

https://github.com/microsoft/ApplicationInsights-Home/blob/master/EndpointSpecs/Schemas/Bond/MetricData.bond

Reply:
> The ingestion service only accepts a single data point at a time, so there is no need for batching.

It is hard to believe that the ingestion service would be able to serve real customer in production without batching.